smarts/env/multi_scenario_env.py

```diff
@@ -282,6 +290,101 @@ def resolve_agent_interface(
     )
 
 
+class _LimitTargetPose(gym.Wrapper):
+    """Uses previous observation to limit the next TargetPose action range."""
+
+    def __init__(self, env: gym.Env):
+        """
+        Args:
+            env (gym.Env): Environment to be wrapped.
+        """
+        super().__init__(env)
+        self._prev_obs: Dict[str, Dict[str, Any]]
+
+    def step(
+        self, action: Dict[str, np.ndarray]
+    ) -> Tuple[
+        Dict[str, Any],
+        Dict[str, float],
+        Dict[str, bool],
+        Dict[str, Dict[str, Any]],
+    ]:
+        """Steps the environment.
+
+        Args:
+            action (Dict[str, Any]): Action for each agent.
+
+        Returns:
+            Tuple[ Dict[str, Any], Dict[str, float], Dict[str, bool], Dict[str, Dict[str, Any]] ]:
+                Observation, reward, done, and info, for each agent is returned.
+        """
+
+        limited_actions: Dict[str, np.ndarray] = {}
+        for agent_name, agent_action in action.items():
+            limited_actions[agent_name] = self._limit(
+                name=agent_name,
+                action=agent_action,
+                prev_coord=self._prev_obs[agent_name].ego_vehicle_state.position[:2],
+            )
+
+        out = self.env.step(limited_actions)
+        self._prev_obs = out[0]
+        return out
+
+    def reset(self, **kwargs) -> Dict[str, Any]:
+        """Resets the environment.
+
+        Returns:
+            Dict[str, Any]: A dictionary of observation for each agent.
+        """
+        obs = self.env.reset(**kwargs)
+        self._prev_obs = obs
+        return obs
+
+    def _limit(
+        self, name: str, action: np.ndarray, prev_coord: np.ndarray
+    ) -> np.ndarray:
+        """Set time delta and limit Euclidean distance travelled in TargetPose action space.
+
+        Args:
+            name (str): Agent's name.
+            action (np.ndarray): Agent's action.
+            prev_coord (np.ndarray): Agent's previous xy coordinate on the map.
+
+        Returns:
+            np.ndarray: Agent's TargetPose action which has fixed time-delta and constrained next xy coordinate.
+        """
+
+        time_delta = 0.1
+        limited_action = np.array(
+            [action[0], action[1], action[2], time_delta], dtype=np.float32
+        )
+        speed_max = 28  # 28m/s = 100.8 km/h. Maximum speed should be >0.
+        dist_max = speed_max * time_delta
+
+        # Set time-delta
+        if not math.isclose(action[3], time_delta, abs_tol=1e-3):
+            logger.warning(
+                f"{name}: Expected time-delta={time_delta}, but got time-delta={action[3]}. "
+                f"Action time-delta automatically changed to {time_delta}."
+            )
```

Comment:
We can skip string operations when logging level is higher than warning by using old c format substitution.

```suggestion
            logger.warning(
                "%s: Expected time-delta=%s, but got time-delta=%s. "
                "Action time-delta automatically changed to %s.",
                name,
                time_delta,
                action[3],
                time_delta 
            )
```